java/vector/src/main/java/org/apache/arrow/vector/ValueVector.java

```diff
@@ -251,4 +251,29 @@
    * @return true if equals, otherwise false.
    */
   boolean equals(int index, ValueVector target, int targetIndex);
+
+  /**
+   * Check whether this vector equals to target vector.
+   * @param target target vector
+   * @return true if they have same type and elements, otherwise false
+   */
+  default boolean equals(ValueVector target) {
+    if (this == target) {
+      return true;
+    }
+    if (target == null || this.getClass() != target.getClass()) {
```

Comment:
Have you thought about semantics of equality between NonNullable and NullableVersions?

Reply:
It reuse equals(int index，ValueVector target, int targetIndex) to compare each value which directly compares its underlying ArrowBuf，do we still need to worry about null value？Add logic to check isNull at index and targetIndex？

Reply:
I thought in some instances we have two different class (E.g. NullableStructValueVector.class  != StructValueVector?

Reply:
Thanks, you are right，haven't thought about that，these two vectors should equals if they have same elements right? Do You have any suggestion on this case? 😀

Reply:
Check Field type?

Reply:
The equals() implementation in BaseFixedWidthVector/BaseVariableWidthVector looks incorrect too - it doesn't check the validity bits. Am i missing something ?

Reply:
@pravindra Right, we should also consider validity buffer, fixed now, thanks a lot.
@emkornfield Now check MinorType instead of getClass I think could solve semantics of equality between NonNullable and NullableVersions

Reply:
regarding the MinorType check - it doesn't work for decimals. eg. "100" with decimal(3,2) is 1.00 but decimal(3,3) is 0.1. but the equals check above would return true.

Reply:
Thanks for pointing, should DecimalVector implement equals and check precision and scale equal first?